cachi2/core/package_managers/yarn/resolver.py

```diff
@@ -344,13 +343,44 @@ def _generate_purl_for_package(package: _ResolvedPackage, project: Project) -> s
     :param project: the project object to resolve the configured registry url and file paths
         for file dependencies.
     """
-    # registry url can be accessed in project.yarnrc
-    # paths for file dependencies are relative to project.source_dir
+    qualifiers = dict()
+    subpath = None
+
+    # package with NpmLocator doesn't need any qualifiers
+
+    if isinstance(package.locator, HttpsLocator):
+        qualifiers["download_url"] = package.locator.url
+        if package.checksum is not None:
+            qualifiers["checksum"] = f"sha512-{package.checksum}"
+
+    elif isinstance(package.locator, WorkspaceLocator):
+        project_path = project.source_dir
+        workspace_path = package.locator.relpath
+
+        repo = get_repo_id(project_path)
+
+        qualifiers["vcs_url"] = repo.as_vcs_url_qualifier()
+        subpath = str(workspace_path)
+
+    elif isinstance(package.locator, (FileLocator, LinkLocator, PortalLocator)):
+        project_path = project.source_dir
+        workspace_path = package.locator.locator.relpath
+        package_path = package.locator.relpath
+
+        repo = get_repo_id(project_path)
+        qualifiers["vcs_url"] = repo.as_vcs_url_qualifier()
+        subpath = str(workspace_path.joinpath(package_path))
```

Comment:
I did a quick test using our test repo as [source](https://github.com/cachito-testing/cachi2-yarn-berry/tree/no-zero-installs). Looking at the results, I got two distinct results for local `strip-ansi:

```sh
# correct, probably derived from the main workspace
pkg:npm/strip-ansi@4.0.0?vcs_url=git%2Bhttps://github.com/cachito-testing/cachi2-yarn-berry%4070515793108df42547d3320c7ea4cd6b6e505c46#external-packages/strip-ansi-4.0.0.tgz

# incorrect, probably derived from packages/then-answer workspace
pkg:npm/strip-ansi@4.0.0?vcs_url=git%2Bhttps://github.com/cachito-testing/cachi2-yarn-berry%4070515793108df42547d3320c7ea4cd6b6e505c46#packages/the-answer/external-packages/strip-ansi-4.0.0.tgz
```

The second one's locator is `file:../../external-packages/strip-ansi-4.0.0.tgz`, so I'm guessing the relative import is not being properly resolved.